File: Tests/parallel_loop_reduction_bitand_general.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.f90"
        INTEGER :: x, y, z !Iterators
        INTEGER,DIMENSION(LOOPCOUNT):: a !Data
        INTEGER :: b
        REAL(8),DIMENSION(16 * LOOPCOUNT):: randoms
        REAL(8) :: false_margin
        INTEGER :: errors = 0
        INTEGER :: temp = 1

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(randoms)
        false_margin = exp(log(.5)/n)
        DO x = 1, LOOPCOUNT
          DO y = 0, 15
            IF (randoms(y * LOOPCOUNT + x) .lt. false_margin) THEN
              temp = 1
              DO z = 0, y
                temp = temp * 2
              END DO
              a(x) = a(x) + temp
            END IF
          END DO
        END DO
        
        b = a(1)

        !$acc data copyin(a(1:LOOPCOUNT))
          !$acc parallel loop reduction(iand:b)
          DO x = 1, LOOPCOUNT
            b = iand(b, a(x))
          END DO
        !$acc end data
 
        temp = a(1)
        DO x = 2, LOOPCOUNT
          temp = iand(a(x), temp)
        END DO
        IF (temp .ne. b) THEN
          WRITE(*, *) temp
          WRITE(*, *) b
          errors = 1
        END IF
        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.f90"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of parallel_loop_reduction_bitand_general"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of parallel_loop_reduction_bitand_general"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM
                                             

File: Tests/parallel_loop_reduction_bitand_general.cpp
#include "acc_testsuite.h"
#ifndef T1
//T1:parallel,loop,reduction,combined-constructs,V:1.0-2.7
int test1(){
    int err = 0;
    srand(SEED);
    n = 10;
    unsigned int * a = (unsigned int *)malloc(n * sizeof(unsigned int));
    real_t false_margin = pow(exp(1), log(.5)/n);
    unsigned int temp = 1;
    unsigned int b;
    unsigned int host_b;

    for (int x = 0; x < n; ++x){
        a[x] = 0;
        for (int y = 0; y < 16; ++y){
            if (rand() / (real_t) RAND_MAX < false_margin){
                for (int z = 0; z < y; ++z){
                    temp *= 2;
                }
                a[x] += temp;
                temp = 1;
            }
        }
    }

    b = a[0];
    host_b = a[0];

    #pragma acc data copyin(a[0:n])
    {
        #pragma acc parallel loop reduction(&:b)
        for (int x = 0; x < n; ++x){
            b = b & a[x];
        }
    }

    for (int x = 1; x < n; ++x){
        host_b = host_b & a[x];
    }
    if (b != host_b){
        err = 1;
    }

    return err;
}
#endif

#ifndef T2
//T2:parallel,reduction,combined-constructs,loop,V:2.7-2.7
int test2(){
    int err = 0;
    srand(SEED);
    unsigned int * a = (unsigned int *)malloc(10 * n * sizeof(int));
    real_t false_margin = pow(exp(1), log(.5)/n);
    unsigned int device[10];
    unsigned int host[10];

    for (int x = 0; x < 10 * n; ++x) {
        a[x] = 0;
        for (int y = 0; y < 16; ++y){
            if (rand() / (real_t)RAND_MAX < false_margin) {
                a[x] += 1<<y;
            }
        }
    }

    for (int x = 0; x < 10; ++x) {
        device[x] = 0;
        host[x] = 0;
        for (int y = 0; y < 16; ++y) {
            device[x] += 1<<y;
            host[x] += 1<<y;
        }
    }

    #pragma acc data copyin(a[0:10*n])
    {
        #pragma acc parallel loop reduction(&:device)
        for (int x = 0; x < 10 * n; ++x) {
            device[x%10] = device[x%10] & a[x];
        }
    }

    for (int x = 0; x < 10 * n; ++x) {
        host[x%10] = host[x%10] & a[x];
    }

    for (int x = 0; x < 10; ++x) {
        if (host[x] != device[x]) {
            err += 1;
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
#ifndef T2
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test2();
    }
    if (failed != 0){
        failcode = failcode + (1 << 1);
    }
#endif
    return failcode;
}
